### ISSUE TITLE
Directly realizing a func with an RDom aborts when realization does not contain RDom bounds.

### ISSUE THREAD

**Issue by alexreinking** [halide/Halide#3883]
If you have a Func (call it `g`) that contains an RDom and you try to realize it directly, then the bounds of that RDom are checked to be within the bounds of the realization (I believe so as not to overwrite the buffer). However, creating an identity Func `h` for `g` and then realizing `h` on those same bounds works as one might expect.

This behavior is confusing especially when calling `realize()` because the Halide API is responsible for allocating a buffer in this case, anyway.

It would be better to allocate a temporary buffer when needed for an output and then copy if the supplied buffer isn't large enough already. The safety check is being performed anyway.

----

The following Halide program demonstrates the behavior:

```
#include <Halide.h>
#include <iostream>
using namespace Halide;

void print_buf2d(Buffer<float> box) {
    int x_lo = box.min(0);
    int x_hi = x_lo + box.extent(0);
    int y_lo = box.min(1);
    int y_hi = y_lo + box.extent(1);
    for (int y = y_lo; y < y_hi; y++) {
        for (int x = x_lo; x < x_hi; x++) {
            std::cout << box(x,y);
        }
        std::cout << "\n";
    }
}

int main() {
    Func f{"f"}, g{"g"}, h{"h"};
    Var x{"x"}, y{"y"};

    f(x, y) = 1.0f;

    RDom r(-10,15,-10,15);
    g(x, y) = 1 + f(x, y);
    g(r.x, r.y) = 1 + g(r.x, r.y);

    h(x,y) = g(x,y);

    Buffer<float> box = h.realize(10, 10);
    print_buf2d(box);

    Buffer<float> box2 = g.realize(10, 10);
    print_buf2d(box2);
}
```

It produces the following output:

```
3333322222
3333322222
3333322222
3333322222
3333322222
2222222222
2222222222
2222222222
2222222222
2222222222
Error:
Output buffer g is accessed at -10, which is before the min (0) in dimension 0
Aborted (core dumped)
```

**Comment by abadams** [comment]
realize's main mode is to evaluate into an existing buffer. It also has a convenient variant that allocates a buffer of the requested size and realizes into it (i.e. you should read `realize(20, 20)` as just `realize(Buffer<float>{20, 20})` followed by returning the new Buffer) I think it would be quite surprising for the returned buffer to not be the size requested. A hard failure seems better to me if it's not possible to realize something at the requested size.

**Comment by abadams** [comment]
Sorry, that's not what you suggested. I think a silent copy is also not ideal because it may be as expensive as the entire rest of the algorithm, and something you may want to schedule. You can explicitly request your described behavior like so: `g.compute_root().in().realize(10, 10);`

**Comment by alexreinking** [comment]
My hang-up with this example is that the way realization is often communicated: as the act of evaluating a Func (which is intuitively some sort of description of a computation). In the [docs for realize()](https://halide-lang.org/docs/class_halide_1_1_func.html#a6360790c98ecd355360fa8ed24b51edc), it says:

> Evaluate this function over some rectangular domain and return the resulting buffer or buffers.

It isn't clear why `h=g` should be more evaluate-able than just `g`. This seems to conflate a scheduling decision implicit in the usage of the "let-Halide-manage-memory" overload of `realize` with the conceptually pure description of the algorithm.

If you were going to do something with `g`, you wouldn't need to interpose a wrapper around it. The treatment of output buffers (and whether or not to make those copies) seems to make more sense as a scheduling decision to me.

**Comment by abadams** [comment]
Copying requires way too much scheduling to be done implicitly (e.g. how should it be done if the source is an OpenCL buffer? Should the result also be a device-side thing only, or should this also imply device->host copy?), but it occurs to me that it would be possible to realize into a larger buffer and return a cropped view of the requested size - not a copy.

**Comment by alexreinking** [comment]
> but it occurs to me that it would be possible to realize into a larger buffer and return a cropped view of the requested size - not a copy.

I would have suggested that if it weren't for

> I think it would be quite surprising for the returned buffer to not be the size requested.

A view of a buffer isn't quite the same as a buffer. I don't know how much value you place on that distinction.

**Comment by alexreinking** [comment]
I'm still not sure why doing that copy when the requested size is too small is an issue. The size check is already running regardless, so there's only additional cost paid in what are now `abort()`-ing cases, right?

**Comment by abadams** [comment]
I think most people doing production work would prefer an error to a silent slowdown. If there needs to be a copy they're going to want to explicitly schedule it.

What are you thinking of when you say a view is not a buffer? Contiguity in memory?

**Comment by alexreinking** [comment]
> What are you thinking of when you say a view is not a buffer? Contiguity in memory?

Yes. You can `memcpy` one but not the other, for instance.

**Comment by abadams** [comment]
non-contiguity worries me far less than an implicit extra allocation and an implicit copy.

**Comment by alexreinking** [comment]
Which existing programs would that extra allocation and copy affect?

**Comment by alexreinking** [comment]
I might not be understanding you.

Where we have, essentially:

```
if output buffer bounds are bad:
  abort()
else
  buffer = run pipeline over bounds
  return buffer
```

I suggested:

```
if output buffer bounds are bad:
  print warning
  tmp = run pipeline over (extended bounds)
  buffer = copy of tmp over bounds
else
  buffer = run pipeline over bounds
return buffer
```

I believe you would prefer:

```
if output buffer bounds are bad:
  buffer = run pipeline over (extended bounds)
else
  buffer = run pipeline over bounds
return view of buffer over bounds
```

I do not know if you prefer this to the existing behavior, however.

**Comment by abadams** [comment]
TL;DR: I think the current behavior is unfriendly, but the proposed behavior is a performance foot-gun.

Mostly a response to your second-last comment - took me a while to write:

No existing programs rely on this behavior of course, but I think existing programmers do. Currently people understand "realize" to mean one allocation of the stated size, with values stored into it directly, and an error if that's not doable. Your proposal would change it to mean usually one allocation of the stated size, but sometimes two where one is a little larger, plus a copy, but not a copy that I get to schedule, so I'd better be ok with just a call to memcpy. I worry that this translates to a slowdown and spike in memory usage that the user doesn't understand.

The view solution would mean that the memory usage is sometimes larger than what I stated (not ideal), and that the memory returned is non-contiguous (also not ideal). I do not know if I prefer this to the existing behavior. I think so? I would want to bounce it off everyone.

The current solution is an unfriendly error message about writing out of bounds on the output buffer.

Some of my concern with the copy is that there isn't just one way to do a memcpy. There are several, and the optimal one depends on the context in which the program runs, and what else has happened and is going to happen with that data. The mere existence of the copy is also probably going to change how I schedule the rest of the program. If there's going to be a copy anyway, then that's a good time to do any final pointwise ops because you have, e.g. idle floating point ALUs. Any time there's a performance trade-off like this to be made, the Halide compiler isn't supposed to just pick something reasonable. It's supposed to get the user to specify the behavior explicitly using the scheduling language (autoschedulers exist above this layer). I also believe should also be able to look at a piece of Halide code and clearly see how many large pieces of memory it's going to allocate by counting the compute_roots (+1 for the output). People working on embedded think about pipelines in terms of the total number of times data goes out to dram, and try to get as much done as possible in between each such instance. Adding an extra one where no useful work gets done is not ok for those users.

**Comment by alexreinking** [comment]
> Currently people understand "realize" to mean one allocation of the stated size, with values stored into it directly, and an error if that's not doable. 

Might I be forgiven for taking away a different understanding of the verb "realize"? I understood it to mean "compute this Func for every point in the requested domain and return a buffer containing the results".

> [...] Your proposal would change it to mean usually one allocation of the stated size, but sometimes two where one is a little larger, plus a copy, but not a copy that I get to schedule, so I'd better be ok with just a call to memcpy. I worry that this translates to a slowdown and spike in memory usage that the user doesn't understand.
>
> The view solution would mean that the memory usage is sometimes larger than what I stated (not ideal), and that the memory returned is non-contiguous (also not ideal). [...]

In either case, I suggested printing a warning to the user explaining that their requested realization has caused an inefficiency. Again, this is only in the context where the user has already asked Halide to manage their memory. The message is in service of doing that better.

> People working on embedded think about pipelines in terms of the total number of times data goes out to dram, and try to get as much done as possible in between each such instance. Adding an extra one where no useful work gets done is not ok for those users.

We already agree that these users' existing code would not be affected by either change. Would a warning that suggested using a supplied-buffer overload and/or calling `in()` not be as good an incentive to fix their programs as the existing `abort()` behavior, only friendlier?

> Some of my concern with the copy is that there isn't just one way to do a memcpy. There are several, and the optimal one depends on the context in which the program runs, and what else has happened and is going to happen with that data. The mere existence of the copy is also probably going to change how I schedule the rest of the program.

I'm not sure what "the rest of the program" is when we're already talking about an output. Are the buffers returned by `realize(int, int)` not always copied to the host? [Line 654 of Pipeline.cpp](https://github.com/halide/Halide/blob/master/src/Pipeline.cpp#L654) leads me to believe they are.

**Comment by abadams** [comment]
We generally view warnings as an antipattern. They should either be errors or the behavior should not be worthy of warning the user. I think returning a cropped view might count as the latter but I would want to get a sufficient subset of @zvookin @dsharletg and @steven-johnson to weigh in.

By rest of the program I mean the last few stages before the end. The existence of a copy in order to fit into a smaller buffer changes how you would want to schedule them. Buffers are not always copied to host when you realize into an existing buffer, but I see that they are when you use the int-args form of realize. I had forgotten that, and I guess that's an consistency argument for making this form of realize more automatic. 

in() and clone_in() are most definitely the schedule under Halide's definition. It affects performance but not correctness, so it preserves the scheduling language guarantee. Your workaround is a change to the algorithm code that obeys the same properties as a scheduling directive, but whenever we see that sort of thing we'd prefer to make it an explicit scheduling directive to advertise to readers that it doesn't affect correctness (e.g. so you can comfortably change the code to do it differently on a different platform and know you're not going to change the output).

**Comment by abadams** [comment]
@shoaibkamil would be another good person to weigh in on the ideal behavior of realize(ints) under these circumstances.

**Comment by alexreinking** [comment]
Might as well also tag @jrk, with whom I was discussing this earlier, though I don't expect him to be available on his honeymoon 🙂
